lib/aiken/rationals.ak

```diff
@@ -0,0 +1,295 @@
+// Rationals
+
+use aiken/math
+
+// Useful for using Rationals in datums/redeemers which cannot have opaque types
+pub type UncheckedRational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Opaque type used to ensure the sign of the Rational is managed strictly
+// in the numerator
+pub opaque type Rational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Converts an UncheckedRational to a Rational
+pub fn check_rational(num: UncheckedRational) -> Rational {
+  let UncheckedRational { numerator, denominator } =
+    num
+  div_int(numerator, denominator)
+}
+
+// Multiplication
+pub fn mul(a: Rational, b: Rational) -> Rational {
+  Rational {
+    numerator: a.numerator * b.numerator,
+    denominator: a.denominator * b.denominator,
+  }
+}
```

Comment:
I've recently learned that pattern matching is more performant than using the field getters. Also this function does not reduce the fraction after the multiplication by dividing by the greatest common divisor.

Here is the mulitplication function I created for my ratio module:
```
/// Multiply two `Rational` types.
pub fn multiply(x: Rational, y: Rational) -> Rational {
  let Rational(x_num,x_den) = x
  let Rational(y_num,y_den) = y
  let new_num =
    x_num * y_num
  let new_den =
    x_den * y_den
  let gcd_ =
    gcd(new_num, new_den)

  Rational { numerator: new_num / gcd_, denominator: new_den / gcd_ }
}
```

Reply:
It is my preference but the function name "mul" is not as clear to me what it does as "multiply". The way I designed my module was to prefer prefacing with `ratio.`. Reading `ratio.multiply` is very clear what the function does.

Reply:
The other arithmetic functions also do not reduce the fraction after each operation. I did some benchmarking and did not see any issues with reducing after each operation.

Reply:
To be precise, some reductions I tested were slightly more expensive after each operation while others were barely noticeable. The performance impact seemed to depend heavily on the fraction being reduced but on net seemed negligible. Another importance factor is how large will the fraction get if you do not reduce after each operation? If the developer needs to decide how often to reduce, if they decide wrong, this could disqualify certain transactions that have several large fractions that must be combined. This can possibly be a security risk if malicious users deliberately try to create large fractions. Given that the performance hit seems negligible on net, I believe it is better to take the slight hit to priorities safety.

Reply:
GCD is strictly more costly than division however memory costs for many uplc arithmetic operations are constant- e.g comparison checks. We should resolve this with actual benchmarks, but I believe conversely to your example, you could build two numbers where their intended use (etc. comparison, for min-fee coin with many decimals) is very cheap, but reducing drives costs towards tx disqualification. We should see some practical examples and how people are actually using rationals, but in my case not reducing is preferable. 

Considering either implementation *could* be problematic, whatever conclusion we reach, I think we should go with one as the default and the other as an additional submodule.